src/Service/ExternalUserService.php

```diff
@@ -0,0 +1,164 @@
+
+    public function __construct(
+        private readonly UserActivationCodeRepository $activationCodeRepository,
+        private readonly EntityManagerInterface $entityManager,
+        private readonly Security $security,
+        private readonly string $hashSalt,
+        private readonly UserRoleTenantRepository $userRoleTenantRepository,
+    ) {}
+
+    public function generatePersonalIdentifierHash(string $personalIdentifier): string
+    {
+        return hash('sha512', $this->hashSalt.$personalIdentifier);
+    }
+
+    /**
+     * @throws CodeGenerationException
+     */
+    public function refreshCode(UserActivationCode $code): UserActivationCode
+    {
+        $code->setCode($this->generateExternalUserCode());
+        $code->setCodeExpire((new \DateTime())->add(new \DateInterval('P2D')));
+        $this->entityManager->flush();
+
+        return $code;
+    }
+
+    /**
+     * @throws ExternalUserCodeException
+     */
+    public function activateExternalUser(string $code): void
+    {
+        /** @var User $user */
+        $user = $this->security->getUser();
+
+        // Make sure user is an external user.
+        if (UserTypeEnum::OIDC_EXTERNAL === !$user->getUserType()) {
+            throw new ExternalUserCodeException('User is not an external type.', 404);
+        }
+
+        $activationCode = $this->activationCodeRepository->findOneBy(['code' => $code]);
+
+        if (null === $activationCode) {
+            throw new ExternalUserCodeException('Activation code not found.', 404);
+        }
+
+        $tenant = $activationCode->getTenant();
+        $roles = $activationCode->getRoles();
+        $displayName = $activationCode->getUsername();
+
+        // The activation code has been used. Remove it.
+        $this->activationCodeRepository->remove($activationCode, true);
+
+        // Set user's fullName if not set.
+        if (empty($user->getFullName()) || self::EXTERNAL_USER_DEFAULT_NAME === $user->getFullName()) {
+            $user->setFullName($displayName);
+        }
+
+        if (null === $user->getEmail()) {
+            $user->setEmail($this->getEmailFromDisplayName($displayName));
+        }
+
+        // Make sure UserRoleTenant does not already exist.
+        $userRoleTenants = $this->userRoleTenantRepository->findBy(['user' => $user, 'tenant' => $tenant]);
+
+        if (count($userRoleTenants) > 0) {
+            throw new ExternalUserCodeException('User already activated for the given tenant.', 400);
+        }
+
+        $userRoleTenant = new UserRoleTenant();
+        $userRoleTenant->setTenant($tenant);
+        $userRoleTenant->setRoles($roles);
+
+        $user->addUserRoleTenant($userRoleTenant);
+
+        $this->entityManager->persist($userRoleTenant);
+
+        $this->entityManager->flush();
+    }
+
+    public function getEmailFromDisplayName(string $displayName): string
+    {
+        $slugged = $this->slugifyDisplayName($displayName);
+
+        return "$slugged@ext";
```

Comment:
I prefer `$slugged.'@ext'` or a `sprintf()` for readbility

Reply:
Fixed